_posts/2021-01-18-smtchecker-and-synthesis-of-external-functions.md

```diff
@@ -0,0 +1,420 @@
+		if (newWeiRaised > cap)
+			newWeiRaised = cap;
+		weiRaised = newWeiRaised;
+	}
+
+	function finalize() public {
+		require(state == State.STARTED);
+		assert(weiRaised < cap);
+		state = State.FINISHED;
+	}
+}
+```
+
+The counterexample is:
+
+```bash
+$ solc crowdsale.sol --model-checker-engine chc
+
+Warning: CHC: Assertion violation happens here.
+Counterexample:
+state = 1, weiRaised = 1, cap = 1
+
+Transaction trace:
+SimpleCrowdsale.constructor()
+State: state = 0, weiRaised = 0, cap = 0
+SimpleCrowdsale.setCap(1)
+State: state = 1, weiRaised = 0, cap = 1
+SimpleCrowdsale.receive()
+State: state = 1, weiRaised = 1, cap = 1
+SimpleCrowdsale.finalize()
+  --> crowdsale.sol:31:3:
+   |
+31 | 		assert(weiRaised < cap);
+   | 		^^^^^^^^^^^^^^^^^^^^^^^
+```
+
+The counterexample shows that after deployment, the assertion violation can be
+reached with 3 transactions. The correct property is `assert(weiRaised <= cap);`.
+
+## External Calls to Unknown Code
+
+External calls most of the time cannot be resolved at compile time.  This
+presents a new challenge in proving contract invariants: they must hold even if
+the contract makes external calls to unknown code that can do anything! This
+includes an unbounded number of potential reentrant calls to the original
+contract, and calls to other contracts.
+
+The `mutex` example below asserts the property that state variable `x` cannot
+be modified inside function `run`. Function `run` does not change `x` directly,
+but it performs an external call to unknown code which can, for example, call
+contract `Mutex` back, trying to change `x` via `Mutex.set`.
+A precise analysis of this case requires not only functional knowledge about
+`Mutex.set`, but also global knowledge about `Mutex`'s behavior in the case of
+any number of reentrant calls, in any order.
+
+```solidity
+//SPDX-License-Identifier: GPL-v3
+pragma solidity ^0.8.0;
+pragma experimental SMTChecker;
+
+interface Unknown {
+	function run() external;
+}
+
+contract Mutex {
+	uint x;
+	bool lock;
+
+	Unknown immutable unknown;
+
+	constructor(Unknown _u) {
+		require(address(_u) != address(0));
+		unknown = _u;
+	}
+
+	modifier mutex {
+		require(!lock);
+		lock = true;
+		_;
+		lock = false;
+	}
+
+	function set(uint _x) mutex public {
+		x = _x;
+	}
+
+	function run() mutex public {
+		uint xPre = x;
+		unknown.run();
+		assert(xPre == x);
+	}
+}
+```
+
+[Spacer](https://spacer.bitbucket.io/),
+the Horn solver used by the CHC engine, automatically infers the inductive
+invariant `lockPre => lockPost && lockPre => (xPost = xPre)` for that specific
+external call, that is, it learns that
+
+- if `lock = true` before the call, it must also be `true` after the call.
+- if `lock = true` before the call, `x` cannot be modified by reentrant calls.
+
+From the local context it knows that `lock = true` before the call.  Therefore,
+it must be true that **no reentrant call** can modify `x`, and the property is
+correct.
```

Comment:
This paragraph is a little confusing to me, don't you show that the property is false just below?

Reply:
This paragraph is still about the previous example, added `in the example above` to clarify